### PR TITLE
93-4 93-5 create a custom view for o auth login and user creation

### PR DESCRIPTION
**Pull request by Chelakhovl**
Created the OAuthTokenObtainPairView, a custom API view that handles the OAuth-based login process.
The view performs the following steps:
OAuth Authentication: Authenticates users using either Google or GitHub OAuth, retrieving the authorization code and exchanging it for an access token.
User Retrieval or Creation: If the user already exists in our system, they are retrieved; if not, a new user is created with the data provided by the OAuth provider.
JWT Token Assignment: Once the user is authenticated, the view assigns them a JWT token for secure access to our platform.
This was implemented by leveraging Django's RefreshToken.for_user() method, which issues the user a new JWT token upon successful OAuth login.
Code References:

OAuthTokenObtainPairView in views.py: Contains the logic for handling Google/GitHub OAuth login, checking for existing users, creating new users as necessary, and assigning JWT tokens.
Task 2: Create a Celery Task to Send a Welcome Email on Successful Registration or Login
Objective:
To configure a Celery task that sends a welcome email notification to the user upon successful registration or login.

Implementation:

Added a new Celery task, send_welcome_email, in tasks.py that sends a welcome email with the user's details.
The task is triggered asynchronously after successful registration or login by calling send_welcome_email.apply_async with the user’s ID as an argument.
Configured the email template to include the user's name and email, making the email more personalized and welcoming.
Set up the Celery task configuration in the settings and ensured that the Celery worker is connected to Redis, allowing for smooth background task processing.
Code References:

send_welcome_email in tasks.py: Contains the logic for sending the welcome email using Django’s send_mail function.
CustomSocialAccountAdapter in adapters.py: This adapter activates users after registration and triggers the send_welcome_email task.
The email configuration and template setup in settings.py ensure that the email is sent from the configured SMTP settings.

### CODE REVIEW

File: forum/users/adapter.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        Returns:
+            User: The saved user instance with an activated status.
+        """
+        user = super().save_user(request, sociallogin, form)
+        user.is_active = True
+        user.save()
+        
+        # Convert user_id to string to pass to Celery task
+        user_id = str(user.user_id)
+        send_welcome_email.apply_async(args=[user_id])
```
Comment (mehalyna) [inline forum/users/adapter.py:38]: Passing `user_id` as a string is unnecessary since the task can handle integers.

File: forum/users/adapter.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            request: The HTTP request object.
+            sociallogin: An instance containing the social account details.
+            form: Optional; form data, if any, associated with the user.
+
+        Returns:
+            User: The saved user instance with an activated status.
+        """
+        user = super().save_user(request, sociallogin, form)
+        user.is_active = True
+        user.save()
```
Comment (mehalyna) [inline forum/users/adapter.py:34]: If the user saving or the task fails, there is no error handling.

File: forum/users/tasks.py [hunk 3]
```diff
@@ -31,6 +36,37 @@ def send_activation_email(user_id, activation_url):
...
+        Sends an info log if the email was sent successfully,
+        or an error log if the user is not found or the email fails to send.
+    """
+    try:
+        user = User.objects.get(user_id=user_id)
+        subject = "Welcome to Our Platform!"
+        message = f"Hi {user.username},\n\nWelcome to our platform! We're glad to have you with us."
+        from_email = settings.EMAIL_HOST_USER
+        recipient_list = [user.email]
+        send_mail(subject, message, from_email, recipient_list, fail_silently=False)
```
Comment (mehalyna) [inline forum/users/tasks.py:67]: You may use Django’s `render_to_string` for creating dynamic email content (like HTML templates), improving maintainability and user experience.

File: forum/users/tasks.py [hunk 4]
```diff
@@ -31,6 +36,37 @@ def send_activation_email(user_id, activation_url):
...
+
+    Args:
+        user_id (str): The UUID of the user to send the welcome email to.
+
+    Logs:
+        Sends an info log if the email was sent successfully,
+        or an error log if the user is not found or the email fails to send.
+    """
+    try:
+        user = User.objects.get(user_id=user_id)
```
Comment (mehalyna) [inline forum/users/tasks.py:62]: Use `select_related()` to avoid multiple queries:
```
user = User.objects.select_related().get(user_id=user_id)
```

**Review by mehalyna — CHANGES_REQUESTED** [review]

**Review by mehalyna — APPROVED** [review]

**Review by gogolmariana — APPROVED** [review]

**Review by k0drin — APPROVED** [review]

**Pull request merged**